extensions/modules/compression/src/main/java/org/exist/xquery/modules/compression/AbstractCompressFunction.java

```diff
@@ -394,23 +394,23 @@ private void compressResource(OutputStream os, DocumentImpl doc, boolean useHier
 		}
 
         final byte[] value;
-		if (doc.getResourceType() == DocumentImpl.XML_FILE) {
-			// xml file
-			Serializer serializer = context.getBroker().getSerializer();
-			serializer.setUser(context.getSubject());
-			serializer.setProperty("omit-xml-declaration", "no");
+        if (doc.getResourceType() == DocumentImpl.XML_FILE) {
+            // xml file
+            Serializer serializer = context.getBroker().getSerializer();
+            serializer.setUser(context.getSubject());
+            serializer.setProperty("omit-xml-declaration", "no");
             getDynamicSerializerOptions(serializer);
             String strDoc = serializer.serialize(doc);
-            value = strDoc.getBytes();            
-		} else if (doc.getResourceType() == DocumentImpl.BINARY_FILE) {
-			// binary file
-            try (final InputStream is = context.getBroker().getBinaryResource((BinaryDocument)doc);
-                 final UnsynchronizedByteArrayOutputStream baos = new UnsynchronizedByteArrayOutputStream(doc.getContentLength() == -1 ? 1024 : (int)doc.getContentLength())) {
+            value = strDoc.getBytes();
+        } else if (doc.getResourceType() == DocumentImpl.BINARY_FILE && doc.getContentLength() > 0) {
+            // binary file
+            try (final InputStream is = context.getBroker().getBinaryResource((BinaryDocument) doc);
+                 final UnsynchronizedByteArrayOutputStream baos = new UnsynchronizedByteArrayOutputStream(doc.getContentLength() == -1 ? 1024 : (int) doc.getContentLength())) {
                 baos.write(is);
                 value = baos.toByteArray();
```

Comment:
What does this produce when `doc.getContentLength() == 0`? I would have thought `byte[0]`

Reply:
I did not check, but the `write()` never ends

Reply:
`doc.getContentLength() == -1` is never true btw

Reply:
at least now `byte[0]` is created :-)